### PR TITLE
Fix Solaris build

### PR DESCRIPTION
**Pull request by sjthomason**
Define __SUNOS on Solaris

### CODE REVIEW

**Comment by sjthomason** [comment]
Initially I noticed that __SUNOS wasn't being defined on Solaris by autoconf but after digging a little deeper I found some additional build issues.  This PR at least allows a build to complete on Solaris but I have a few questions regarding the projects stance on a few items:
- Can u_intX_t be replaced with standard uintX_t or would a typedef (or CPP define) be preferable on platforms that require it?
- Is is preferable to relay on autoconf to define __SUNOS or the more standard
```
#if (defined(__SVR4) && defined(__sun))
```

**Comment by wdoekes** [comment]
Thanks for the PR!

> Can u_intX_t be replaced with standard uintX_t or would a typedef (or CPP define) be preferable on platforms that require it?

Yes. It'd be better if they were replaced by the standardized types. No one has taken the time to fix this though. Any patches are welcome.

> Is is preferable to relay on autoconf to define __SUNOS or the more standard [...]

Standards are better :)

That being said, I believe there was another fix for the missing le16toh():
https://github.com/SIPp/sipp/commits/fix/3.5/add-missing-le16toh-on-old-glibc-211
Does that work for you?

File: include/endianshim.h [hunk 1]
```diff
@@ -13,6 +13,15 @@
...
 #endif
 
+#ifdef __SUNOS
+#include <sys/byteorder.h>
+#ifdef _BIG_ENDIAN
+#define le16toh(x) BSWAP_16(x)
+#else
+#define le16toh(x) (x)
+#endif
+#endif
```
Comment (wdoekes) [inline include/endianshim.h]: I merged the endian changes just now. Do those fix this bit for you?

File: src/socket.cpp [hunk 2]
```diff
@@ -1685,7 +1685,11 @@ void SIPpSocket::sipp_sctp_peer_params()
         memset(&peerparam, 0, sizeof(peerparam));
 
         sockaddr* addresses;
+#ifdef __SUNOS
+        int addresscount = sctp_getpaddrs(ss_fd, 0, (void **) &addresses);
+#else
         int addresscount = sctp_getpaddrs(ss_fd, 0, &addresses);
```
Comment (wdoekes) [inline src/socket.cpp]: How about casting to `(struct sockaddr**)` instead. Then you can drop the `#ifdef`.

Comment (sjthomason) [inline src/socket.cpp]: Solaris uses a different prototype unfortunately:
int sctp_getpaddrs(int sock, sctp_assoc_t id, void **addrs)

Without the cast to the cast to ```(void **)``` g++ throws an error:
```error: invalid conversion from ‘sockaddr**’ to ‘void**’```

File: include/endianshim.h (hunk 1 above)
Comment (sjthomason) [inline include/endianshim.h]: Unfortunately Solaris needs a different header.  I updated the PR with your recent changes and used the builtin BSWAP_16 macro on this platform.

Thanks!

**Comment by wdoekes** [comment]
Thanks! I cherrypicked/adapted the changes so they could be merge upwards from 3.5. Could not test, but I think it should works.

**Comment by sjthomason** [comment]
Hi @wdoekes,  it appears a necessary change didn't make it to master:
```
diff --git a/src/send_packets.c b/src/send_packets.c
index 12838fb..5610e3f 100644
--- a/src/send_packets.c
+++ b/src/send_packets.c
@@ -39,9 +39,9 @@
 #include <stdlib.h>
 #include <stdbool.h>
 #include <stdint.h>
-#include <netinet/udp.h>
 #include <netinet/in.h>
 #include <netinet/ip6.h>
+#include <netinet/udp.h>
 #include <errno.h>
 #include <string.h>
 #include <fcntl.h>
```

The include order is significant on some platforms:
```/usr/include/netinet/udp.h:26:2: error: unknown type name ‘in_port_t’```

**Comment by wdoekes** [comment]
Drat! Sorry about that =)

Fixed now.

**Pull request closed without merging**